File: leetcode/easy/reverse_linkedlist.ts
import { ListNode } from "../common/common";

function reverseList(head: ListNode | null): ListNode | null {
  let reversed = null;
  let current = head;

  while (current) {
    let nextTemp = current.next;
    current.next = reversed;
    reversed = current;
    current = nextTemp;
  }

  return reversed;
}

File: leetcode/easy/reverse_linkedlist.py
# Given the head of a singly linked list, reverse the list, and return the reversed list.
# Input: head = [1,2,3,4,5]
# Output: [5,4,3,2,1]
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        prev = None
        curr = head
        while curr:
            nextListNode = curr.next
            curr.next = prev
            prev = curr

            curr = nextListNode

        return prev
